oxide-auth/src/code_grant/extensions/pkce.rs

```diff
@@ -151,24 +192,32 @@ impl Method {
         }
     }
 
-    fn from_encoded(encoded: Cow<str>) -> Result<Method, ()> {
-        // TODO: avoid allocation in case of borrow and invalid.
-        let mut encoded = encoded.into_owned();
-        match encoded.pop() {
-            None => Err(()),
-            Some('p') => Ok(Method::Plain(encoded)),
-            Some('S') => Ok(Method::Sha256(encoded)),
-            _ => Err(()),
+    fn from_encoded(encoded: Cow<str>) -> Result<Method, PkceError> {
+        match encoded.chars().last() {
+            Some('p') => {
+                let mut string = avoid_alloc_cow_str_to_string(encoded);
+                string.pop();
+                Ok(Method::Plain(string))
+            }
+            Some('S') => {
+                let mut string = avoid_alloc_cow_str_to_string(encoded);
+                string.pop();
+                Ok(Method::Sha256(string))
+            }
+            _ => Err(PkceError::InvalidMethod(
+                "No/invalid encoding character indicator at end of string, expected `p` or `S`"
+                    .to_string(),
```

Comment:
Explanatory string should be part of the `Display` impl, not a string formatting.